### ISSUE TITLE
Need proper solution for SetPostResolveHook

### ISSUE THREAD

**Issue by mathewc** [Azure/azure-webjobs-sdk#1013]
We have several bindings that currently rely on this behavior, including 2 of our own as well as an external 3rd party binding (Bot Framework).

We recently made this functionality internal, but we ourselves continue to use it because it is useful. However that has broken external customers. We shouldn't have our own bindings using any substantial internal infrastructure like this - if we need it, it is likely useful to eternal binding authors as well.

To unblock the Bot guys (and avoid breaking changes for them) I put in a temporary hack using reflection to invoke this internal method. We need to remove that.

**Comment by MikeStall** [comment]
For background, this hook does 2 very tactical things, mainly around supporting a fixed set of backwards-compat scenarios that shouldn't apply to extensions: 
1. Preserve existing rich Dashboard support in the core bindings. Our new bindings don’t customize this. We can transfer this responsibility to the binding rules and find good ways to customizations for other bindings – but that was a lower priority and so not done yet.  
2. Support cases where an attribute is split across multiple attributes, ie BlobStorage and StorageAccount.   From the Function's perspective, this is an anti-pattern since it breaks the 1:1 mapping between attributes and Function.json. New extensions shouldn't do this.  Instead, new extensions should have all relevant properties within a single attribute. 

Both of these motivations are limited in scope and don’t apply to new extensions, so new extensions shouldn’t need the hook.

**Comment by mathewc** [comment]
If we can do away with the hooks completely then great. This issue is mainly about making any changes we need to remove the private reflection hack in the Bot extension. If that means getting them to take the breaking change, then we might not have to give them any new API. However it's tricky because we made the API public, they built on it, then we did a take back.

**Comment by mathewc** [comment]
I think this can be closed now, right @MikeStall ?
